location.go

```diff
@@ -141,23 +149,39 @@ func newRawLocationData(r *http.Request) *rawLocationData {
 
 func (fl *rawLocationData) locationData() (*locationData, error) {
 	lat, ok := parseFloat(fl.Latitude)
-	if !ok {
+	if !ok || lat < -maxLatitude || lat > maxLatitude {
 		return nil, newClientError("Invalid latitude")
 	}
 	lon, ok := parseFloat(fl.Longitude)
-	if !ok {
+	if !ok || lon < -maxLongitude || lon > maxLongitude {
 		return nil, newClientError("Invalid longitude")
 	}
 	ts, ok := parseTimestamp(fl.Timestamp)
 	if !ok {
 		return nil, newClientError("Invalid timestamp")
 	}
 	alt, ok := parseFloat(fl.Altitude)
-	if !ok && fl.Altitude != "" {
+	if !ok && fl.Altitude == "" {
+		ok = true // This field is optional, so can be left empty
+	} else {
+		// A value has been set.  Check if it's a sane one.
+		if alt < minAltitude || alt > maxAltitude {
+			ok = false
+		}
+	}
+	if !ok {
 		return nil, newClientError("Invalid altitude")
 	}
 	pre, ok := parseFloat(fl.Precision)
-	if !ok && fl.Precision != "" {
+	if !ok && fl.Precision == "" {
+		ok = true // This field is optional, so can be left empty
+	} else {
+		// A value has been set.  Check if it's a sane one.
+		if pre < minPrecision || pre > maxPrecision {
+			ok = false
+		}
+	}
+	if !ok {
```

Comment:
this function would be a lot cleaner if the validation was moved into a separate method which could do early returns with an appropriate error if a problem was found

Reply:
Yeah I knew it was getting kinda messy.

So do you mean each of those 2 last checks (altitude and precision) validate in a separate little function?

Reply:
I can do this.  Or I can rewrite the ParseFloat function I guess?

`	pre := 0.0
	if fl.Precision != "" {
		// A value has been set.  Check if it's a sane one.
		pre, ok := parseFloat(fl.Precision)
		if !ok || pre < minPrecision || pre > maxPrecision {
			return nil, newClientError("Invalid precision")
		}
	}
`

Reply:
Oh that didn't format well...

Reply:
`	pre := 0.0

	if fl.Precision != "" {
		// A value has been set.  Check if it's a sane one.
		pre, ok := parseFloat(fl.Precision)
		if !ok || pre < minPrecision || pre > maxPrecision {
			return nil, newClientError("Invalid precision")
		}
	}
`

Reply:
And to help the user I could do this I suppose:

`return nil, newClientError(fmt.Sprintf("Invalid altitude. Should be between %2.2f and %2.2f", minAltitude, maxAltitude)
`

Reply:
> ```
> pre := 0.0
> if fl.Precision != "" {
> 	// A value has been set.  Check if it's a sane one.
> 	pre, ok := parseFloat(fl.Precision)
> 	if !ok || pre < minPrecision || pre > maxPrecision {
> 		return nil, newClientError("Invalid precision")
> 	}
> }
> ```

This is good, but do this in a separate validation method which called from the top of `locationData`.